### ISSUE TITLE
.Net Core 2.2

### ISSUE THREAD

**Issue by jcoburn** [loic-sharma/BaGet#439]
Latest installation 0.20 requires .Net Core 2.2 but that is already pas end of support.  Plans for .Net Core 3.1 LTS or .Net Core 2.1 LTS support?

**Comment by MarkZither** [comment]
I have been having a look at this, and noticed that it is already raised by @loic-sharma on #285 

so far i have been following the steps in [migrate 2.2 to 3.0](https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-3.1&tabs=visual-studio) and [migrate 3.0 to 3.1](https://docs.microsoft.com/en-us/aspnet/core/migration/30-to-31?view=aspnetcore-3.1&tabs=visual-studio).

- [x] Update the project file and remove/add relevant package references
- [x] Change `BaGet.Core.Server` to use `Microsoft.NET.Sdk.Web` and target `netcoreapp3.1` for MVC references
- [ ] Changes to remove `AddJsonOptions DateTimeZoneHandling = DateTimeZoneHandling.Utc` to be compatible with System.Text.Json (removed it, need to understand the impact)
- [x] Change from `IHostingEnvironment` to `IWebHostEnvironment` 
- [ ] Migrate from deprecated `UseMvc` to `UseRouting` and `UseEndpoints` (having issues finding the controllers in `BaGet.Core.Server`. Checking [Share controllers, views, Razor Pages and more with Application Parts](https://docs.microsoft.com/en-us/aspnet/core/mvc/advanced/app-parts?view=aspnetcore-3.1) and [.NET core 3 route failure with separate projects](https://github.com/dotnet/aspnetcore/issues/14952)
- [x] Fix a duplicate route that is throwing and exception for `SymbolDownloadRouteName` as also done in [wagnerhsu's fork](https://github.com/wagnerhsu/tools-BaGet/commit/45da688a74c7c066dbc1a93f6cedc2561484b36e)
- [ ] Look at the new EF warnings `warn: Microsoft.EntityFrameworkCore.Model.Validation[10620]
      The property 'Authors' on entity type 'Package' is a collection or enumeration type with a value converter but with no value comparer. Set a value comparer to ensure the collection/enumeration elements are compared correctly.
warn: Microsoft.EntityFrameworkCore.Model.Validation[10620]
      The property 'Tags' on entity type 'Package' is a collection or enumeration type with a value converter but with no value comparer. Set a value comparer to ensure the collection/enumeration elements are compared correctly.`

I have it running but as the controllers and routes are not working it doesn't do very much. Hopefully i will be able to submit a PR soon.

**Comment by MarkZither** [comment]
Seems i got a bit carried away and now it is probably too big for an unsolicited PR.
I have updated the comment above to split it into required and optional work, happy to submit smaller PRs if they are welcome. but i think it has the side effect of dropping support for 4.6.1 so might need further thought.

**Comment by loic-sharma** [comment]
Hey @MarkZither, thank you for getting the ball rolling on the ASP.NET Core 3 upgrade, I appreciate the help! I'd be happy to accept a pull request for the required changes. Let's wait on the optional changes for now to keep the pull request manageable.

> it has the side effect of dropping support for 4.6.1 so might need further thought.

I'm okay with this, but let's keep the `BaGet.Protocol` project on `netstandard2.0`.

> Seems i got a bit carried away and now it is probably too big for an unsolicited PR.

Haha that happens 😊

**Comment by loic-sharma** [comment]
@MarkZither I started working on this using your tips: https://github.com/loic-sharma/BaGet/compare/update-aspnet3

Current status:
1. My controllers are currently broken. It seems you ran into a similar issue, have you figured this out? I'll keep digging on this front
1. The latest PostgreSQL driver does not like running migrations synchronously and throws an exception. As a work around, I now move the migrations in `Program.cs` asynchronously
1. ASP.NET Core 3 found some dependency injection issues. I cleaned those up

**Comment by MarkZither** [comment]
hi @loic-sharma re the problem with the controllers, I solved it by setting the applicationkey
`.UseSetting(WebHostDefaults.ApplicationKey, Assembly.GetExecutingAssembly().GetName().Name);` in program.cs.

I will revert my routing changes for now, in favour of `services.AddMvc(option => option.EnableEndpointRouting = false)`.

**Comment by loic-sharma** [comment]
@MarkZither Thanks for the info! BaGet does add the `BaGet.Core.Server` assembly as an application part (see [this](https://github.com/loic-sharma/BaGet/blob/dcbd51df2247651e5341c394de691c829b4f60f0/src/BaGet.Core.Server/Extensions/IServiceCollectionExtensions.cs#L18)). I believe the issue is because of [the async suffix removal from controller action names](https://github.com/loic-sharma/BaGet/blob/dcbd51df2247651e5341c394de691c829b4f60f0/src/BaGet.Core.Server/Extensions/IServiceCollectionExtensions.cs#L18) in ASP.NET Core 3, as many of the routes had the `Async` suffix (for example [this](https://github.com/loic-sharma/BaGet/blob/dcbd51df2247651e5341c394de691c829b4f60f0/src/BaGet.Core.Server/Extensions/IRouteBuilderExtensions.cs#L14)).

I'm not sure why `.UseSetting(WebHostDefaults.ApplicationKey, Assembly.GetExecutingAssembly().GetName().Name);` fixes the issue, but I don't think you can use this solution on generic hosts. Considering that the web host is now deprecated, I'd prefer a solution that does need the `UseSetting` trick.

**Comment by MarkZither** [comment]
oh yes, sorry, the async suffix caught me out as well.

**Comment by loic-sharma** [comment]
Ok I think I've figured out everything for the ASP.NET Core 3.0 migration on this branch: https://github.com/loic-sharma/BaGet/compare/update-aspnet3

I found some bugs that I think I'll fix on the current master branch before I start the migration. I'll look at the 3.1 migration tomorrow. Thanks again @MarkZither, you have been very helpful!

**Comment by loic-sharma** [comment]
The latest release runs on .NET Core 3.1. Please let me know if you have any additional issues or feedback!
